mclf/curves/smooth_projective_curves.py

```diff
@@ -1265,3 +1271,69 @@ def e_f_of_valuation(v):
     if hasattr(v, "_approximation"):
         v = v._approximation
     return (v.E(), v.F())
+
+
+def extension_of_finite_field(K, n):
+    r""" Return a field extension of this finite field of degree n.
+
+    INPUT:
+
+    - ``K`` -- a finite field
+    - ``n`` -- a positive integer
+
+    OUTPUT: a field extension of `K` of degree `n`.
+
+    This function is useful if `K` is constructed as an explicit extension
+    `K = K_0[x]/(f)`; then ``K.extension(n)`` is not implemented.
+
+    """
+    assert K.is_field()
+    assert K.is_finite()
+    q = K.order()
+    R = PolynomialRing(K, 'z')
+    z = R.gen()
+    # we look for a small number e dividing q^n-1 but not q-1
+    e = min([d for d in (q**n-1).divisors() if not d.divides(q-1)])
+    F = (z**e-1).factor()
+    f = [g for g, e in F if g.degree() == n][0]
+    # this is very inefficient!
+    return K.extension(f, 'z'+str(e))
+
+
+def make_finite_field(k):
+    r""" Return the finite field isomorphic to this field.
+
+    INPUT:
+
+    - ``k`` -- a finite field
+
+    OUTPUT: a pair `(k_1,\phi)` where `k_1` is a 'true' finite field
+    and `\phi` is an isomorphism from `k` to `k_1`.
+
+    This function is useful when `k` is constructed as a tower of extensions
+    with a finite field as a base field.
+
+    """
+    from sage.categories.finite_fields import FiniteFields
+
+    assert k.is_field()
+    assert k.is_finite()
+    if k in FiniteFields:
```

Comment:
Is this ever true? `FiniteFields` is a method that returns the category. It should probably read `k in FiniteFields()`?

Then again, that's true for a tower whose top extension is a quotient so it's probably the wrong check. I'm confused because it seemed to work for you.

Reply:
Of course, ``FiniteFields`` is a typo and was intended as ``FiniteFields()``. 

I find it very hard in general to find out what kind of object one is dealing with (use `` in category``, or ``type`` or ``hasattr``?). It is also very hard to find out from the documentation.

Reply:
Yes that's truly confusing, I don't have anything smart to say about this except that Sage is really lacking that regard.